### PR TITLE
refactor: avoid unstable let_chains

### PR DESCRIPTION
**Pull request by strager**
The let_chains Rust feature is unstable, preventing Oxc from using a stable Rust compiler. Refactor the code to avoid let_chains.

#626

### CODE REVIEW

File: crates/oxc_minifier/src/compressor/fold.rs [hunk 1]
```diff
@@ -458,23 +454,30 @@ impl<'a> Compressor<'a> {
...
-                        && (left.operator, right.operator) == (UnaryOperator::Typeof, UnaryOperator::Typeof)
-                        && let Expression::Identifier(left) = &left.argument
-                        && let Expression::Identifier(right) = &right.argument
-                        && left.name == right.name
+                    if let (Expression::UnaryExpression(left), Expression::UnaryExpression(right)) =
+                        (left_expr, right_expr)
                     {
-                        // Special case, typeof a == typeof a is always true.
-                        return Tri::True;
+                        if (left.operator, right.operator)
```
Comment (strager) [inline crates/oxc_minifier/src/compressor/fold.rs:469]: This is identical to line 380. Perhaps it should be factored into a function. If it's in a function, early returns might make more sense. Thoughts @Boshen?

**Comment by Boshen** [comment]
Thank you in advance @strager! I wrote so much `let_chains` 😢 Apparently 5 years and still no stabilization `https://github.com/rust-lang/rust/issues/53667` ...

File: crates/oxc_minifier/src/compressor/mod.rs [hunk 2]
```diff
@@ -195,9 +206,12 @@ impl<'a> Compressor<'a> {
 
     /// Transforms `undefined` => `void 0`
     fn compress_undefined<'b>(&mut self, expr: &'b mut Expression<'a>) -> bool {
-        if let Expression::Identifier(ident) = expr
-        && ident.name == "undefined"
-        && self.semantic.symbols().is_global_reference(ident.reference_id.clone().into_inner()) {
+        let Expression::Identifier(ident) = expr else {
+            return false;
+        };
```
Comment (strager) [inline crates/oxc_minifier/src/compressor/mod.rs:211]: @Boshen Are guard clauses like this acceptable?

**Comment by Boshen** [comment]
> [CLA assistant check](https://cla-assistant.io/web-infra-dev/oxc?pullRequest=628) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/web-infra-dev/oxc?pullRequest=628) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/web-infra-dev/oxc?pullRequest=628) it.

Sorry this requires a CLA :-(

File: crates/oxc_minifier/tests/terser/mod.rs [hunk 3]
```diff
@@ -125,13 +138,14 @@ impl TestCase {
...
         for object_property in &object_expr.properties {
-            if let ObjectPropertyKind::ObjectProperty(property) = object_property
-                && let Some(name) = property.key.static_name() {
-                match name.as_str() {
-                    "drop_debugger" => {
-                        options.drop_debugger = Self::get_boolean(&property.value);
+            if let ObjectPropertyKind::ObjectProperty(property) = object_property {
+                if let Some(name) = property.key.static_name() {
+                    match name.as_str() {
+                        "drop_debugger" => {
```
Comment (strager) [inline crates/oxc_minifier/tests/terser/mod.rs:144]:
```rust
                if property.key.static_name().is_some_and(|name| name == "drop_debugger") {
                    options.drop_debugger = Self::get_boolean(&property.value);
                }
```

File: crates/oxc_semantic/src/binder.rs [hunk 4]
```diff
@@ -57,13 +57,16 @@ impl<'a> Binder for VariableDeclarator<'a> {
...
 impl<'a> Binder for Class<'a> {
     fn bind(&self, builder: &mut SemanticBuilder) {
-        if let Some(ident) = &self.id && !self.modifiers.contains(ModifierKind::Declare) {
-            builder.declare_symbol(
-                ident.span,
-                &ident.name,
-                SymbolFlags::Class,
-                SymbolFlags::ClassExcludes,
-            );
+        match &self.id {
```
Comment (strager) [inline crates/oxc_semantic/src/binder.rs:60]: I think a guard clause is better here.

**Review by strager — COMMENTED** [review]
@Boshen Here's a preview of a world without let_chains.

**Comment by Boshen** [comment]
@strager You are free to refactor some of the code. I don't have a preferred style as long as CI passes 😁

File: crates/oxc_hir/src/hir_util.rs [hunk 5]
```diff
@@ -410,11 +410,14 @@ pub fn get_boolean_value(expr: &Expression) -> Option<bool> {
...
-            } else {
-                None
+            if let Some(quasi) = template_literal.quasis.get(0) {
+                if quasi.tail {
+                    return Some(
+                        quasi.value.cooked.as_ref().map_or(false, |cooked| !cooked.is_empty()),
+                    );
+                }
             }
+            None
```
Comment (magic-akari) [inline crates/oxc_hir/src/hir_util.rs:420]:
```suggestion
            template_literal
                .quasis
                .get(0)
                .filter(|quasi| quasi.tail)
                .map(|quasi| quasi.value.cooked.as_ref().map_or(false, |cooked| !cooked.is_empty()))
```

File: crates/oxc_hir/src/hir_util.rs [hunk 6]
```diff
@@ -502,11 +505,16 @@ fn get_string_value<'a>(expr: &'a Expression) -> Option<Cow<'a, str>> {
...
+            if let Some(quasi) = template_literal.quasis.get(0) {
+                if quasi.tail {
+                    return quasi
+                        .value
+                        .cooked
+                        .as_ref()
+                        .map(|cooked| Cow::Borrowed(cooked.as_str()));
+                }
             }
+            None
```
Comment (magic-akari) [inline crates/oxc_hir/src/hir_util.rs:517]:
```suggestion
            template_literal
                .quasis
                .get(0)
                .filter(|quasi| quasi.tail)
                .and_then(|quasi| quasi.value.cooked.as_ref())
                .map(|cooked| Cow::Borrowed(cooked.as_str()))
```

**Comment by strager** [comment]
Note to maintainers: I prefer that this pull request be rebased or merged as-is, not squashed.

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Thank you both @strager and @magic-akari! One step closer to stable Rust.

**Pull request merged**